File: go-api/internal/fraud/vectorize.go
package fraud

import (
	"time"
)

func Vectorize(payload FraudScoreRequest, config *Config) [14]float32 {
	normalization := config.Normalization

	amount := clamp(payload.Transaction.Amount / normalization.MaxAmount)

	installments := clamp(
		float64(payload.Transaction.Installments) / normalization.MaxInstallments,
	)

	amountVsAvg := 0.0
	if payload.Customer.AvgAmount > 0 {
		amountVsAvg = clamp(
			(payload.Transaction.Amount / payload.Customer.AvgAmount) /
				normalization.AmountVsAvgRatio,
		)
	}

	requestedAt, _ := time.Parse(time.RFC3339, payload.Transaction.RequestedAt)

	hourOfDay := clamp(float64(requestedAt.UTC().Hour()) / 23)

	dayOfWeek := getNormalizedDayOfWeek(requestedAt)

	minutesSinceLastTx := float64(-1)
	kmFromLastTx := float64(-1)

	if payload.LastTransaction != nil {
		lastTimestamp, err := time.Parse(time.RFC3339, payload.LastTransaction.Timestamp)

		if err == nil {
			diffMinutes := requestedAt.Sub(lastTimestamp).Minutes()
			minutesSinceLastTx = clamp(diffMinutes / normalization.MaxMinutes)
		}

		kmFromLastTx = clamp(
			payload.LastTransaction.KmFromCurrent / normalization.MaxKm,
		)
	}

	kmFromHome := clamp(payload.Terminal.KmFromHome / normalization.MaxKm)

	txCount24h := clamp(
		float64(payload.Customer.TxCount24h) / normalization.MaxTxCount24h,
	)

	isOnline := float64(0)
	if payload.Terminal.IsOnline {
		isOnline = 1
	}

	cardPresent := float64(0)
	if payload.Terminal.CardPresent {
		cardPresent = 1
	}

	unknownMerchant := float64(1)
	if contains(payload.Customer.KnownMerchants, payload.Merchant.ID) {
		unknownMerchant = 0
	}

	mccRisk := 0.5
	if value, exists := config.MccRisk[payload.Merchant.MCC]; exists {
		mccRisk = clamp(value)
	}

	merchantAvgAmount := clamp(
		payload.Merchant.AvgAmount / normalization.MaxMerchantAvgAmount,
	)

	return [14]float32{
		float32(amount),
		float32(installments),
		float32(amountVsAvg),
		float32(hourOfDay),
		float32(dayOfWeek),
		float32(minutesSinceLastTx),
		float32(kmFromLastTx),
		float32(kmFromHome),
		float32(txCount24h),
		float32(isOnline),
		float32(cardPresent),
		float32(unknownMerchant),
		float32(mccRisk),
		float32(merchantAvgAmount),
	}
}

func clamp(value float64) float64 {
	if value < 0 {
		return 0
	}

	if value > 1 {
		return 1
	}

	return value
}

func getNormalizedDayOfWeek(date time.Time) float64 {
	weekday := date.UTC().Weekday()

	// Go:
	// Sunday = 0
	// Monday = 1
	// ...
	// Saturday = 6
	//
	// Rinha:
	// Monday = 0
	// Tuesday = 1
	// ...
	// Sunday = 6

	if weekday == time.Sunday {
		return 1
	}

	return float64(weekday-1) / 6
}

func contains(items []string, value string) bool {
	for _, item := range items {
		if item == value {
			return true
		}
	}

	return false
}

File: src/utils/vectorize.js
import fs from "node:fs";
import { getNormalizedDayOfWeek } from "../utils/get-normalized-day-of-week.js";
import { limitar } from "../utils/clamp.js";

const normalizationFile = fs.readFileSync(
  "./src/datasets/normalization.json",
  "utf-8",
);
const normalization = JSON.parse(normalizationFile);

const merchantFile = fs.readFileSync("./src/datasets/mcc_risk.json", "utf-8");
const mcc = JSON.parse(merchantFile);

export function vectorize(data) {
  const normalizatedData = {
    amount: limitar(data.transaction.amount / normalization.max_amount),
    installments: limitar(
      data.transaction.installments / normalization.max_installments,
    ),
    amount_vs_avg: limitar(
      data.transaction.amount /
        data.customer.avg_amount /
        normalization.amount_vs_avg_ratio,
    ),
    hour_of_day: new Date(data.transaction.requested_at).getUTCHours() / 23,
    day_of_week: getNormalizedDayOfWeek(data.transaction.requested_at),
    minutes_since_last_tx:
      data.last_transaction === null
        ? -1
        : limitar(
            (new Date(data.transaction.requested_at).getTime() -
              new Date(data.last_transaction.timestamp).getTime()) /
              1000 /
              60 /
              normalization.max_minutes,
          ),
    km_from_last_tx:
      data.last_transaction === null
        ? -1
        : limitar(data.last_transaction.km_from_current / normalization.max_km),
    km_from_home: limitar(data.terminal.km_from_home / normalization.max_km),
    tx_count_24h: limitar(
      data.customer.tx_count_24h / normalization.max_tx_count_24h,
    ),
    is_online: data.terminal.is_online ? 1 : 0,
    card_present: data.terminal.card_present ? 1 : 0,
    unknown_merchant: data.customer.known_merchants.includes(data.merchant.id)
      ? 0
      : 1,
    mcc_risk: mcc[data.merchant.mcc] ?? 0.5,
    merchant_avg_amount: limitar(
      data.merchant.avg_amount / normalization.max_merchant_avg_amount,
    ),
  };

  return [
    normalizatedData.amount,
    normalizatedData.installments,
    normalizatedData.amount_vs_avg,
    normalizatedData.hour_of_day,
    normalizatedData.day_of_week,
    normalizatedData.minutes_since_last_tx,
    normalizatedData.km_from_last_tx,
    normalizatedData.km_from_home,
    normalizatedData.tx_count_24h,
    normalizatedData.is_online,
    normalizatedData.card_present,
    normalizatedData.unknown_merchant,
    normalizatedData.mcc_risk,
    normalizatedData.merchant_avg_amount,
  ];
}
